onnx/funconnx/op_run.py

```diff
@@ -0,0 +1,508 @@
+        The default answer is `False`.
+        """
+        return False
+
+    def __str__(self) -> str:
+        atts = [self.__class__.__name__ + "(", f"    op_type={self.onnx_node.op_type}"]
+        for k, v in sorted(self.__dict__.items()):
+            if k in {"desc", "onnx_node"}:
+                continue
+            if "a" <= k[0] <= "z" and k[-1] != "_":
+                atts.append(f"    {k}={v},")
+        atts.append(")")
+        return "\n".join(atts)
+
+    @abstractmethod
+    def _run(self, *args, **kwargs):  # type: ignore
+        """
+        Should be overwritten.
+
+        :param args: operator inputs
+        :param kwargs: optional inputs
+        :return: outputs
+        """
+        raise NotImplementedError(
+            f"Method '_run' must be overwritten for operator {self.__class__.__name__!r}."
+        )
+
+    def run(self, *args, linked_attributes=None, **kwargs):  # type: ignore
+        """
+        Calls method ``_run``, catches exceptions,
+        displays a longer error message.
+        """
+        if self.has_linked_attribute and linked_attributes is None:
+            raise ValueError(
+                f"This node {type(self)} has linked attributes but None are given in parameter 'linked_attributes'."
+            )
+        if not self.has_linked_attribute and linked_attributes is not None:
+            raise ValueError(
+                f"This node {type(self)} has no linked attribute but some are given in parameter 'linked_attributes' {set(linked_attributes)}."
+            )
+        overridden_attributes = {}
+        if self.has_linked_attribute:
+            if linked_attributes is None:
+                raise AttributeError(
+                    f"One attribute is linked but no linked value is provided, "
+                    f"in class {type(self)}."
+                )
+            for att in self.attributes_names_:
+                v = getattr(self, att)
+                if isinstance(v, RefAttrName):
+                    if v.name not in linked_attributes:
+                        raise ValueError(
+                            f"Unable to find a value for linked attribute {att!r} in {linked_attributes!r} "
+                            f"in node {type(self)}."
+                        )
+                    overridden_attributes[att] = linked_attributes[v.name]
+
+        self._log("-- begin %s.run(%d inputs)", self.__class__.__name__, len(args))
+        kwargs = kwargs.copy()
+        for att in self.attributes_names_:
+            if att in overridden_attributes or att in kwargs:
+                continue
+            if not hasattr(self, att):
+                raise NameError(
+                    f"Attribute {att!r} is missing in operator {self.__class__.__name__!r}."
+                )
+            kwargs[att] = getattr(self, att)
+        try:
+            if len(overridden_attributes) > 0:
+                res = self._run(*args, **overridden_attributes, **kwargs)
```

Comment:
Bit unclear what happens if a name appears in both overridden_attributes and kwargs. May be easier to understand it if it is made explicit in line 355-356 by choosing the correct one there.

Reply:
I clarified.